src/eos-updater-apply.c

```diff
@@ -73,41 +77,592 @@ get_test_osname (void)
+   * flatpaks to install into a pointer array. Parse out the OSTree ref
+   * and then parse the FlatpakRemoteRef */
+  g_autofree gchar *contents = NULL;
+  g_auto(GStrv) lines = NULL;
+  g_autoptr(GPtrArray) flatpaks = NULL;
+  GStrv lines_iter = NULL;
+
+  if (!g_file_get_contents (g_file_get_path (file), &contents, NULL, error))
+    return NULL;
+
+  lines = g_strsplit (contents, "\n", -1);
+  flatpaks = g_ptr_array_new_full (g_strv_length (lines), g_object_unref);
+
+  for (lines_iter = lines; *lines_iter; ++lines_iter)
+    {
+      g_autofree gchar *remote = NULL;
+      g_autofree gchar *ref = NULL;
+      FlatpakRef *flatpak_ref = NULL;
+
+      if (!ostree_parse_refspec (*lines_iter, &remote, &ref, error))
+        return NULL;
+
+      flatpak_ref = flatpak_ref_parse (ref, error);
+
+      if (!flatpak_ref)
+        return NULL;
+
+      g_ptr_array_add (flatpaks,
+                       g_object_new (FLATPAK_TYPE_REMOTE_REF,
+                                     "remote-name", remote,
+                                     "name", flatpak_ref_get_name (flatpak_ref),
+                                     "kind", flatpak_ref_get_kind (flatpak_ref),
+                                     NULL));
+    }
+
+  return g_steal_pointer (&flatpaks);
+}
+
+typedef gpointer (*PtrCopyFunc) (gpointer src);
+
+static GPtrArray *
+concat_ptrarrays (GPtrArray *left,
+                  GPtrArray *right,
+                  PtrCopyFunc copy_func,
+                  GDestroyNotify destroy_func)
+{
+  gsize left_size = left ? left->len : 0;
+  gsize right_size = right ? right->len: 0;
+  GPtrArray *ptr_array = g_ptr_array_new_full (left_size + right_size, destroy_func);
+  gsize i;
+
+  for (i = 0; i < left_size; ++i)
+    g_ptr_array_add (ptr_array, copy_func (g_ptr_array_index (left, i)));
+
+  for (i = 0; i < right_size; ++i)
+    g_ptr_array_add (ptr_array, copy_func (g_ptr_array_index (right, i)));
+
+  return ptr_array;
+}
+
+static GPtrArray *
+list_flatpak_refs_from_d_directory (GFile        *directory,
+                                    GCancellable *cancellable,
+                                    GError       **error)
+{
+  g_autofree gchar *autoinstall_flatpak_contents = NULL;
+  g_autofree gchar *autoinstall_flatpak_file_path = NULL;
+  g_auto(GStrv) autoinstall_flatpak_lines = NULL;
+  g_autoptr(GFileEnumerator) autoinstall_d_enumerator = NULL;
+  g_autoptr(GPtrArray) autoinstall_flatpaks = NULL;
+
+  /* Repository checked out, read all files in orer and build up a list
+   * of flatpaks to auto-install */
+  autoinstall_d_enumerator = g_file_enumerate_children (directory,
+                                                        "standard::*",
+                                                        G_FILE_QUERY_INFO_NONE,
+                                                        cancellable,
+                                                        error);
+
+  if (!autoinstall_d_enumerator)
+    return NULL;
+
+  while (TRUE)
```

Comment:
Any way to refactor this as a `for` to make the termination condition more explicit?

Reply:
Apparently using an infinite loop is the correct pattern: https://developer.gnome.org/gio/stable/GFileEnumerator.html#g-file-enumerator-iterate - I imagine that is because `g_file_enumerator_iterate` returns `FALSE` on failure and sets the out-param `file` to `NULL` when it is done.

Reply:
I can't think of a way to do this which would not be very awkward. Putting it in a for loop isn't really possible because the termination condition is an outparam and you also need to store the return value of the function to ensure that it didn't fail, the best that I can think of would be something like this:

    GFile *file = NULL;
    gboolean ok = TRUE;

    if (!g_file_enumerator_iterate (enum, NULL, &file, cancellable, error))
      return NULL;

    for (; file != NULL; ok = g_file_enumerator_iterate ...)
      if (!file)
        break;

      if (!some_other_condition)
        return NULL;
    }

And I'm not even certain that will work. Putting it in a `while` loop has the same problem as well.

Reply:
OK, leave it as it is. Thanks for thinking about this :)

Reply:
No worries - I agree that the existing API is awkward, but it is what it is I suppose.